### PR TITLE
Mitigate race conditions, towards supporting invariant named value refs

### PR DESCRIPTION
**Pull request by agrrr3**
* make NamedRef aware if it is a independent lookup
** in this case the value ref might not (yet) be registered
** else the value ref was just registered before

* mitigate calculation of invariances
** if a value ref is not yet registered, try to sleep and retry 4 times before giving up (up to a second whole)
** giving up defaults to "not invariant", which means calculation optimisations cant be applied
** this will be rare, hopefully only in cases where parse order is wrong

* warn if performance-degradation happens

* fix a missing value ref in master

* use std::future inside of NamedValueRefManager to wait for parse to finish whenever trying to read from the registries
** use this when calculating checksums, putting the NamedValueRefManager last in order to let the other managers also ensure that parsing is finished

* fix and add missing SpecialsManager to server/client checksumming 

* use locking when waiting for parse to finish - there were multiple race conditions

### CODE REVIEW

**Comment by geoffthemedio** [comment]
````
/home/travis/build/freeorion/freeorion/universe/NamedValueRefManager.cpp: In instantiation of ‘void ValueRef::NamedRef<T>::SetTopLevelContent(const string&) [with T = Visibility; std::__cxx11::string = std::__cxx11::basic_string<char>]’:
/home/travis/build/freeorion/freeorion/universe/NamedValueRefManager.cpp:354:17:   required from here
/home/travis/build/freeorion/freeorion/universe/NamedValueRefManager.cpp:306:36: error: cannot bind non-const lvalue reference of type ‘const char*&’ to an rvalue of type ‘const char*’
                       << ") on a " << ( content_name == "THERE_IS_NO_TOP_LEVEL_CONTENT" ? "top-level" : "named-in-the-middle" )
````

**Comment by o01eg** [comment]
I've build it and run tests. Still got errors:

```
18:30:54.448797 {0x00007f604b06e000} [warn] ai : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 215410 != client 2208271

19:03:22.099748 {0x00007f3c2a675440} [warn] test : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2208271 != client 215410
```

**Comment by agrrr3** [comment]
@o01eg now there is a relevant commit

**Comment by o01eg** [comment]
6f0d7c6
```
21:07:09.141251 {0x00007ff179ff6440} [warn] test : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 4319172 != client 2326311

21:09:20.629952 {0x00007f089cf0b000} [warn] ai : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2326311 != client 4319172
```

**Comment by agrrr3** [comment]
> [6f0d7c6](https://github.com/freeorion/freeorion/commit/6f0d7c6c3e4d2824afde5b54108ed39427a4ef1d)
> 
> ```
> 21:07:09.141251 {0x00007ff179ff6440} [warn] test : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 4319172 != client 2326311
> 
> 21:09:20.629952 {0x00007f089cf0b000} [warn] ai : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2326311 != client 4319172
> ```

Cant reproduce. Ran "make test" six times or so.

@o01eg Could you grep the log for the phrase `Giving up` ? If that is there i could understand mismatchs happen.

**Comment by o01eg** [comment]
7752d54

Different launches:

```
08:24:16.784560 {0x00007f585325d700} [error] server : NamedValueRefManager.cpp:231 : NamedRef<T>::NamedRefInitInvariants() still could not find value ref. Giving up.
```

```
08:27:09.184134 {0x00007f801adec700} [error] test : NamedValueRefManager.cpp:231 : NamedRef<T>::NamedRefInitInvariants() still could not find value ref. Giving up.
08:27:14.043954 {0x00007f5f03f55700} [error] server : NamedValueRefManager.cpp:231 : NamedRef<T>::NamedRefInitInvariants() still could not find value ref. Giving up.
08:27:14.655956 {0x00007f5f03f55700} [error] server : NamedValueRefManager.cpp:231 : NamedRef<T>::NamedRefInitInvariants() still could not find value ref. Giving up.
```

File: universe/NamedValueRefManager.cpp [hunk 1]
```diff
@@ -198,35 +201,78 @@ template void RegisterValueRef(std::string name, std::unique_ptr<ValueRef::Value
...
+{
+    if (m_invariants_initialized)
+        return true;
+    auto* vref = GetValueRef();
+    if (!vref) {
+        std::chrono::milliseconds msecs(200);
+        InfoLogger() << "NamedRef<T>::NamedRefInitInvariants() could not find value ref, will sleep a bit and retry.";
+        std::this_thread::sleep_for(msecs);
+        vref = GetValueRef();
+        for (int = 2; i < 5; i++) {
```
Comment (o01eg) [inline universe/NamedValueRefManager.cpp]: `i` was missed. Also multiple tries looks fishy. Can we have it more determined?

**Comment by o01eg** [comment]
d0c65bf

```
18:30:04.319907 {0x00007f3265076700} [error] test : NamedValueRefManager.cpp:231 : NamedRef<T>::NamedRefInitInvariants() still could not find value ref after trying 2 times. Giving up.
18:30:18.545277 {0x00007fd698044700} [error] server : NamedValueRefManager.cpp:231 : NamedRef<T>::NamedRefInitInvariants() still could not find value ref after trying 2 times. Giving up.
18:30:46.326426 {0x00007f9b30726700} [error] ai : NamedValueRefManager.cpp:231 : NamedRef<T>::NamedRefInitInvariants() still could not find value ref after trying 2 times. Giving up.
```

```
18:34:11.645082 {0x00007f583aa65700} [error] test : NamedValueRefManager.cpp:231 : NamedRef<T>::NamedRefInitInvariants() still could not find value ref after trying 2 times. Giving up.
18:34:16.583208 {0x00007f43ebae4700} [error] server : NamedValueRefManager.cpp:231 : NamedRef<T>::NamedRefInitInvariants() still could not find value ref after trying 2 times. Giving up.
18:34:17.183411 {0x00007f43ebae4700} [error] server : NamedValueRefManager.cpp:231 : NamedRef<T>::NamedRefInitInvariants() still could not find value ref after trying 3 times. Giving up.
```

```
18:35:51.293010 {0x00007f9b5f0dc700} [error] server : NamedValueRefManager.cpp:231 : NamedRef<T>::NamedRefInitInvariants() still could not find value ref after trying 2 times. Giving up.
18:35:51.893220 {0x00007f9b5f0dc700} [error] server : NamedValueRefManager.cpp:231 : NamedRef<T>::NamedRefInitInvariants() still could not find value ref after trying 3 times. Giving up.
18:36:08.890477 {0x00007f896f339700} [error] ai : NamedValueRefManager.cpp:231 : NamedRef<T>::NamedRefInitInvariants() still could not find value ref after trying 2 times. Giving up.
```

File: universe/NamedValueRefManager.cpp (hunk 1 above)
Comment (agrrr3) [reply universe/NamedValueRefManager.cpp]: > `i` was missed. Also multiple tries looks fishy. Can we have it more determined?

if i can figure out a way to do the intialisation after parsing we are basically done
my problem is that the AppInterface/ServerApp::StartBackgroundParsing (which knows when parsing the contents is finished) does not have references to the created value refs. Probably could add a function like InitialiseAfterParsingContents to those managers which would relay that to their content type objects

the question i want now to answer, if your mismatches stop if no "Giving up" happens.

**Comment by o01eg** [comment]
5f96a51

```
22:09:24.171197 {0x00007f6fefc29700} [error] test : NamedValueRefManager.cpp:231 : NamedRef<T>::NamedRefInitInvariants() still could not find value ref after trying 2 times. Giving up.
22:09:28.891662 {0x00007fbd34a2b700} [error] server : NamedValueRefManager.cpp:231 : NamedRef<T>::NamedRefInitInvariants() still could not find value ref after trying 2 times. Giving up.
22:09:29.491921 {0x00007fbd34a2b700} [error] server : NamedValueRefManager.cpp:231 : NamedRef<T>::NamedRefInitInvariants() still could not find value ref after trying 3 times. Giving up.
22:09:44.970509 {0x00007f6ff506e440} [warn] test : ClientApp.cpp:168 : Checksum received from server does not match client checksum.
22:09:44.970550 {0x00007f6ff506e440} [warn] test : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 8320321 != client 2326311
22:09:47.873791 {0x00007f8481702700} [error] ai : NamedValueRefManager.cpp:231 : NamedRef<T>::NamedRefInitInvariants() still could not find value ref after trying 2 times. Giving up.
```

```
22:07:21.589763 {0x00007f3199bd5700} [error] server : NamedValueRefManager.cpp:231 : NamedRef<T>::NamedRefInitInvariants() still could not find value ref after trying 2 times. Giving up.
22:07:34.883464 {0x00007f216147a440} [info] test : ClientApp.cpp:165 : Checksum received from server matches client checksum.
22:07:50.247289 {0x00007fd8ef3d8000} [warn] ai : ClientApp.cpp:168 : Checksum received from server does not match client checksum.
22:07:50.247333 {0x00007fd8ef3d8000} [warn] ai : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2326311 != client 2263250
22:07:50.780265 {0x00007fbc1552a000} [info] ai : ClientApp.cpp:165 : Checksum received from server matches client checksum.
```

```
22:06:31.583568 {0x00007fced7f8f700} [error] server : NamedValueRefManager.cpp:231 : NamedRef<T>::NamedRefInitInvariants() still could not find value ref after trying 2 times. Giving up.
22:06:44.756316 {0x00007efc8684f440} [info] test : ClientApp.cpp:165 : Checksum received from server matches client checksum.
22:07:00.775415 {0x00007f4d9cdc4000} [warn] ai : ClientApp.cpp:168 : Checksum received from server does not match client checksum.
22:07:00.775443 {0x00007f4d9cdc4000} [warn] ai : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2326311 != client 4319172
22:07:00.816494 {0x00007f2423888000} [warn] ai : ClientApp.cpp:168 : Checksum received from server does not match client checksum.
22:07:00.816595 {0x00007f2423888000} [warn] ai : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2326311 != client 4211038
```

**Comment by agrrr3** [comment]
This is also thought to mitigate #3220 

> my guess is that the invariants are part the checksums. but these are subject to race conditions without a bigger refactoring.
if that is so a reasonable workaround would be to take the invariants out of the checksums as those are derived anyway.
>
> the clean solution would be to initialise the invariants after parsing is done. another solution would be to initialise invariants on first use - introducing a "m_invariants_initialized" flag into all valuerefs

**Comment by agrrr3** [comment]
> 5f96a51

@o01eg the log output was wrong, now there was also a missing value ref in master, i fixed those and squashed/rebased. 

also note with this high sleep the smoketest was reeeeally taking ages when the value ref was missing

**Comment by o01eg** [comment]
803e37717

```
07:06:10.309661 {0x00007fc3fb973740} [info] test : ClientApp.cpp:165 : Checksum received from server matches client checksum.
07:06:11.102952 {0x00007f8b5df27480} [warn] ai : ClientApp.cpp:168 : Checksum received from server does not match client checksum.
07:06:11.102969 {0x00007f8b5df27480} [warn] ai : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2877675 != client 859547
07:06:11.147091 {0x00007f1b9b051480} [warn] ai : ClientApp.cpp:168 : Checksum received from server does not match client checksum.
07:06:11.147097 {0x00007f1b9b051480} [warn] ai : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2877675 != client 8861689
```

**Comment by agrrr3** [comment]
> [803e377](https://github.com/freeorion/freeorion/commit/803e37717112bd41897164c00b5b110bc163e299)

thank you. ~~Any "Giving up" in the log files?~~

(i dont expect any; i think there is another checksum error  ~~on your installation I cant replicate on mine~~)

edit1: can replicate, no "Giving up"

File: universe/NamedValueRefManager.cpp (hunk 1 above)
Comment (agrrr3) [reply universe/NamedValueRefManager.cpp]: so the answer is, yes mismatches still occur even if all and i found that the value refs from xenophobic.macros did not register on the AI client (WTF?). For getting feedback I removed the NamedReal value refs from the macro file.

Anyway this means the remaining mismatches are unrelated to this PR.

**Comment by o01eg** [comment]
3c9c573

```
23:01:31.140614 {0x00007f708a2fb740} [info] test : ClientApp.cpp:165 : Checksum received from server matches client checksum.
23:01:31.586557 {0x00007f8603aa5480} [warn] ai : ClientApp.cpp:168 : Checksum received from server does not match client checksum.
23:01:31.586566 {0x00007f8603aa5480} [warn] ai : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2863741 != client 4803567
23:01:31.630782 {0x00007f2af550d480} [warn] ai : ClientApp.cpp:168 : Checksum received from server does not match client checksum.
23:01:31.630794 {0x00007f2af550d480} [warn] ai : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2863741 != client 4803567
```

```
23:01:25.439841 {0x00007f99eb5e0740} [info] test : ClientApp.cpp:165 : Checksum received from server matches client checksum.
23:01:25.988487 {0x00007fde7b53e480} [info] ai : ClientApp.cpp:165 : Checksum received from server matches client checksum.
23:01:26.078542 {0x00007f9926dc8480} [warn] ai : ClientApp.cpp:168 : Checksum received from server does not match client checksum.
23:01:26.078548 {0x00007f9926dc8480} [warn] ai : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2863741 != client 859547
```

File: default/scripting/policies/CONFEDERATION.focs.txt [hunk 2]
```diff
@@ -18,7 +18,7 @@ Policy
                 Not ResourceSupplyConnected empire = LocalCandidate.Owner condition = Source
             ]
             effects = SetTargetInfluence value = Value
-                        + (NamedRealLookup name = "SUPPLY_DISCONNECTED_INFLUENCE_FLAT")
+                        + (NamedReal name = "PLC_CONFEDERATION_SUPPLY_DISCONNECTED_INFLUENCE_FLAT" value = 1)
```
Comment (Oberlus) [inline default/scripting/policies/CONFEDERATION.focs.txt]: This was a typo from me (still WIP), I have a SUPPLY_DISCONNECTED_INFLUENCE_FLAT and SUPPLY_DISCONNECTED_INFLUENCE_MALUS.

There are two places where this is (should) be referenced (with the same name, obviously), here and in default/scripting/species/common/influence.macros. The point is Confederation nullifies the influence malus from supply-disconnection, so if at some point someone changes supply-disconnection malus then Confederation has to be updated. Hence the reference value in named_values.macros.

File: default/stringtables/en.txt [hunk 3]
```diff
@@ -11409,7 +11409,7 @@ PLC_CONFEDERATION_DESC
 '''Planets of the empire are organized in a confederacy of independent systems. Colonies do not suffer reduced influence if not connected to the capital, and get a boost to ground troops, but are slightly less stable in general.
 
 Increases max [[metertype METER_TROOPS]] by [[value PLC_CONFEDERATION_MAX_TROOPS_FLAT]] but reduces [[metertype METER_TARGET_HAPPINESS]] ([[value PLC_CONFEDERATION_TARGET_HAPPINESS_FLAT]]) in all planets.
-Increases [[metertype METER_TARGET_INFLUENCE]] on colonies disconnected from the capital by [[value SUPPLY_DISCONNECTED_INFLUENCE_FLAT]].'''
+Increases [[metertype METER_TARGET_INFLUENCE]] on colonies disconnected from the capital by [[value PLC_CONFEDERATION_SUPPLY_DISCONNECTED_INFLUENCE_FLAT]].'''
```
Comment (Oberlus) [inline default/stringtables/en.txt:11412]: As per above comment, I would revert this (or I will change it in subsequent PRs).

File: default/scripting/policies/CONFEDERATION.focs.txt (hunk 2 above)
Comment (agrrr3) [reply default/scripting/policies/CONFEDERATION.focs.txt]: not sure what you propose - I guess using SUPPLY_DISCONNECTED_INFLUENCE_MALUS ?

Comment (agrrr3) [reply default/scripting/policies/CONFEDERATION.focs.txt]: it is not helpful if the FLAT is defined name-in-the-middle and the MALUS in named_values

or one could use a macro for the flat value and use that in both NamedReal definitions

Comment (Oberlus) [reply default/scripting/policies/CONFEDERATION.focs.txt]: I'm not sure what are "name-in-the-middle".
The idea is to have SUPPLY_DISCONNECTED_INFLUENCE_FLAT declared in named.values and use NamedRealLookup in CONFEDERATION and influence.macros. That's the current state in my next PR (not yet finished and pushed).

Comment (Oberlus) [reply default/scripting/policies/CONFEDERATION.focs.txt]: > or one could use a macro for the flat value and use that in both NamedReal definitions

That's better (than what I had in mind).

Comment (agrrr3) [reply default/scripting/policies/CONFEDERATION.focs.txt]: i will implement that then

**Comment by agrrr3** [comment]
@o01eg could you try with the current  a9f8427  ?

can we get this merged so invariants work 99.9% of the time?

**Comment by o01eg** [comment]
a9f8427

```
11:04:54.904392 {0x00007fc5d986d740} [info] test : ClientApp.cpp:165 : Checksum received from server matches client checksum.
11:04:55.472801 {0x00007f464d92b480} [warn] ai : ClientApp.cpp:168 : Checksum received from server does not match client checksum.
11:04:55.472806 {0x00007f464d92b480} [warn] ai : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2863741 != client 859547
11:04:55.627846 {0x00007ff109e36480} [warn] ai : ClientApp.cpp:168 : Checksum received from server does not match client checksum.
11:04:55.627853 {0x00007ff109e36480} [warn] ai : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2863741 != client 2848339
```

```
11:06:08.741251 {0x00007f1988616740} [info] test : ClientApp.cpp:165 : Checksum received from server matches client checksum.
11:06:09.458233 {0x00007fb42be59480} [warn] ai : ClientApp.cpp:168 : Checksum received from server does not match client checksum.
11:06:09.458247 {0x00007fb42be59480} [warn] ai : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2863741 != client 8861689
11:06:09.605157 {0x00007f3da8047480} [warn] ai : ClientApp.cpp:168 : Checksum received from server does not match client checksum.
11:06:09.605163 {0x00007f3da8047480} [warn] ai : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2863741 != client 2781025
```

```
11:06:47.274226 {0x00007f76b9945740} [info] test : ClientApp.cpp:165 : Checksum received from server matches client checksum.
11:06:47.728999 {0x00007f5e74605480} [warn] ai : ClientApp.cpp:168 : Checksum received from server does not match client checksum.
11:06:47.729006 {0x00007f5e74605480} [warn] ai : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2863741 != client 859547
11:06:47.882249 {0x00007fe930650480} [info] ai : ClientApp.cpp:165 : Checksum received from server matches client checksum.
```

**Comment by agrrr3** [comment]
@o01eg funky, dont know what is different in your setup. is it possible that the AI connects to the server/receives the server checksum message before parsing is finished? i guess i will provide another PR with some added debug statements for debugging your setup.

**Comment by o01eg** [comment]
Yes, it possible parsing wasn't finished when AI accept checksum.

**Comment by agrrr3** [comment]
> Yes, it possible parsing wasn't finished when AI accept checksum.

Ok, that could explain the error pattern (only that NamedValueRefManager is showing up). The getchecksum calls on the other managers block on the pending parse of the specific type, parsing named_values.focs.txt blocks StartBackgroundParsing. If StartBackgroundParsing is started concurrently to handling server messages, there is a race condition.

**Comment by agrrr3** [comment]
@o01eg could you try with the current?

**Comment by o01eg** [comment]
@agrrr3 It mostly works but 1 of 10 I get mismatch:

```
07:35:17.262196 {0x00007fd9cf7aa740} [warn] test : ClientApp.cpp:168 : Checksum received from server does not match client checksum.
07:35:17.262205 {0x00007fd9cf7aa740} [warn] test : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2863741 != client 6784314
07:35:17.868402 {0x00007f3cec558480} [info] ai : ClientApp.cpp:165 : Checksum received from server matches client checksum.
07:35:17.883773 {0x00007ff037d18480} [info] ai : ClientApp.cpp:165 : Checksum received from server matches client checksum.
```

**Comment by agrrr3** [comment]
hm i start running out of ideas. maybe one of the other parses is not finished. but then also that checksum should be wrong. maybe there is a parse using named value refs which is not checked in Universe::Checksum?
```
$ for manager in BuildingTypeManager Encyclopedia  FieldTypeManager ShipHullManager ShipPartManager PredefinedShipDesignManager SpeciesManager TechManager ; do find -iname "*.cpp" -exec grep ${manager}::GetCheckSum -A 1 {} \+ ; done
./universe/BuildingType.cpp:unsigned int BuildingTypeManager::GetCheckSum() const {
./universe/BuildingType.cpp-    CheckPendingBuildingTypes();
./universe/Encyclopedia.cpp:unsigned int Encyclopedia::GetCheckSum() const {
./universe/Encyclopedia.cpp-    unsigned int retval{0};
./universe/FieldType.cpp:unsigned int FieldTypeManager::GetCheckSum() const {
./universe/FieldType.cpp-    CheckPendingFieldTypes();
./universe/ShipHull.cpp:unsigned int ShipHullManager::GetCheckSum() const {
./universe/ShipHull.cpp-    CheckPendingShipHulls();
./universe/ShipPart.cpp:unsigned int ShipPartManager::GetCheckSum() const {
./universe/ShipPart.cpp-    CheckPendingShipParts();
./universe/ShipDesign.cpp:unsigned int PredefinedShipDesignManager::GetCheckSum() const {
./universe/ShipDesign.cpp-    CheckPendingDesignsTypes();
./universe/Species.cpp:unsigned int SpeciesManager::GetCheckSum() const {
./universe/Species.cpp-    CheckPendingSpeciesTypes();
./universe/Tech.cpp:unsigned int TechManager::GetCheckSum() const {
./universe/Tech.cpp-    CheckPendingTechs();
```

**Comment by agrrr3** [comment]
The only places of Named*Lookup refvars are used in specials and ship parts.

And actually the SpecialsManager is not used in CheckSumContent (in Universe). That could explain it.

**Comment by agrrr3** [comment]
@o01eg please retry

**Comment by o01eg** [comment]
Mostly works but sometimes I got

```
21:12:03.782897 {0x00007f5bb1909500} [info] test : ClientApp.cpp:165 : Checksum received from server matches client checksum.
21:12:09.536353 {0x00007f2dcc982c00} [warn] ai : ClientApp.cpp:168 : Checksum received from server does not match client checksum.
21:12:09.536476 {0x00007f2dcc982c00} [warn] ai : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2863741 != client 839144
21:12:09.536539 {0x00007f2dcc982c00} [warn] ai : ClientApp.cpp:173 : Checksum for SpeciesManager on server 41223 != client 0
```

```
22:04:17.030951 {0x00007f2a160ed500} [warn] test : ClientApp.cpp:168 : Checksum received from server does not match client checksum.
22:04:17.030986 {0x00007f2a160ed500} [warn] test : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2863741 != client 839144
22:04:17.031007 {0x00007f2a160ed500} [warn] test : ClientApp.cpp:173 : Checksum for SpeciesManager on server 41223 != client 0
22:04:20.109465 {0x00007f182d941c00} [info] ai : ClientApp.cpp:165 : Checksum received from server matches client checksum.
22:04:20.321270 {0x00007f9177fd6c00} [info] ai : ClientApp.cpp:165 : Checksum received from server matches client checksum.
```

**Comment by agrrr3** [comment]
> Mostly works but sometimes I got
> 
> ```
>...
> 22:04:17.031007 {0x00007f2a160ed500} [warn] test : ClientApp.cpp:173 : Checksum for SpeciesManager on server 41223 != client 0
>...
> ```

This looks like the SpeciesManager sometimes fails to compute the right checksum - checksum 0 probably means that there was no species (yet). `SpeciesManager::GetChecksum` correctly calls `CheckPendingSpeciesTypes()`. That one looks a bit different than the others, but it calls `Pending::SwapPending` which contains the waiting for the parse to finish.

That the NamedValueRefManager fails if the SpeciesManager fails in this way is expected (that would be the signal if the NamedRealLookup in a species was not yet parsed).

**Comment by o01eg** [comment]
Log of test when it happened: [15.log.zip](https://github.com/freeorion/freeorion/files/5544862/15.log.zip)

**Comment by agrrr3** [comment]
> Log of test when it happened: [15.log.zip](https://github.com/freeorion/freeorion/files/5544862/15.log.zip)

Thanks for that. Parsing of the Lookup valref seems to work both on server and client (I see the SetTopLevelContent call). 
I see a single entry (not sure about where this happens; the thread(?) is called test):
```
10:49:49.256646 {0x00007fe83d216740} [error] test : Pending.h:77 : Parsing of "specials" failed with error: std::future_error: No associated state
```

i think that indicates trying to get() from the future a second time. hunt continues.

Also some strange stuff i see in the log: some files are parsed once, twice, three times, even four times. There is a human client and two AI clients. Two of the clients report correct checksum.

**Comment by o01eg** [comment]
> There is a human client and two AI clients. Two of the clients report correct checksum.

The server also parses them so files should be parsed three times or four because AI, server and test-client parses same files and additionally the server parses its own files.

**Comment by o01eg** [comment]
Still got errors with b2a37bc

```
20:03:21.991732 {0x00007f6f25598740} [warn] test : ClientApp.cpp:168 : Checksum received from server does not match client checksum.
20:03:21.991752 {0x00007f6f25598740} [warn] test : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2863741 != client 6784314
20:03:21.991767 {0x00007f6f25598740} [warn] test : ClientApp.cpp:173 : Checksum for SpecialsManager on server 896323 != client 0
20:03:22.677975 {0x00007fcf5bc6b480} [info] ai : ClientApp.cpp:165 : Checksum received from server matches client checksum.
20:03:22.861403 {0x00007fe1bacbc480} [info] ai : ClientApp.cpp:165 : Checksum received from server matches client checksum.
```

[14.zip](https://github.com/freeorion/freeorion/files/5561550/14.zip)

**Comment by o01eg** [comment]
Or even

```
20:02:06.564390 {0x00007f9589e94740} [info] test : ClientApp.cpp:165 : Checksum received from server matches client checksum.
20:02:07.094365 {0x00007f9ece103480} [warn] ai : ClientApp.cpp:168 : Checksum received from server does not match client checksum.
20:02:07.094375 {0x00007f9ece103480} [warn] ai : ClientApp.cpp:173 : Checksum for NamedValueRefManager on server 2863741 != client 2754882
20:02:07.094383 {0x00007f9ece103480} [warn] ai : ClientApp.cpp:173 : Checksum for SpecialsManager on server 896323 != client 0
20:02:07.094390 {0x00007f9ece103480} [warn] ai : ClientApp.cpp:173 : Checksum for SpeciesManager on server 41223 != client 0
20:02:07.191477 {0x00007feb65974480} [info] ai : ClientApp.cpp:165 : Checksum received from server matches client checksum.
```

[10.zip](https://github.com/freeorion/freeorion/files/5561557/10.zip)

**Comment by Oberlus** [comment]
Would the suggestion in the following post help with the race conditions problem?
https://www.freeorion.org/forum/viewtopic.php?p=104357#p104357

**Comment by agrrr3** [comment]
> Would the suggestion in the following post help with the race conditions problem?
> https://www.freeorion.org/forum/viewtopic.php?p=104357#p104357

The current race conditions problem (checksum sometimes is calculated before parsing swapped in the results) is not completely understood, I do not see a connection to what you wrote.

About the named values approach: Going the macro route has its own pitfalls and named values debugging is almost done, so i do not think that is a good idea to change course.

I see one "easy" workaround - synchronize all calls waiting for parse results, this should ensure that the values are swapped in. This could lead to unecessary waiting times though.

**Comment by Oberlus** [comment]
@agrrr3 I know you've invested a lot of work into the named values reference manager and its debugging.
It does not seem that complications are close to be solved.
Plus performance issues that could happen on each turn, maybe many times per turn, is a serious issue.

It's never too late to change course, if it is towards a better destination.
I've tried to explain better my suggestion in the linked thread, that would for sure involve no performance or concurrency issues at all. If you don't see any drawback from the sketched approach, I think you can prefer to use it and abandon (fully or in part) the named values approach.

**Comment by agrrr3** [comment]
> @agrrr3 I know you've invested a lot of work into the named values reference manager and its debugging.
> It does not seem that complications are close to be solved.
> Plus performance issues that could happen on each turn, maybe many times per turn, is a serious issue.

dont get me wrong. you do not understand the issues involved here.

**Comment by o01eg** [comment]
fcbd304 works, I've tried it 40 times.

**Comment by agrrr3** [comment]
> [fcbd304](https://github.com/freeorion/freeorion/commit/fcbd304fbcc44df8e72d0cedd704d0e42e929161) works, I've tried it 40 times.

@o01eg Yes i thought it would - i got out the big gun (synchronize all wait calls). Did you experience slowdowns?

**Comment by agrrr3** [comment]
@geoffthemedio @Vezzra can we disable that CI mac build as we know it is failing? it only creates noise and eats resources

**Comment by o01eg** [comment]
@agrrr3 I fixed it recently

**Comment by agrrr3** [comment]
> @agrrr3 I fixed it recently

oh cool :) 
thanks a ton

**Comment by Vezzra** [comment]
> can we disable that CI mac build as we know it is failing?

Fine with me.

**Comment by o01eg** [comment]
@agrrr3 Could you try rebase this PR first?

**Comment by agrrr3** [comment]
> @agrrr3 Could you try rebase this PR first?

Sure

**Comment by agrrr3** [comment]
i think i confused the build system by rebasing and squashing some stuff

**Comment by agrrr3** [comment]
Ok, CI sais the macOS build is failing because it times out after 50minutes.

**Comment by agrrr3** [comment]
this solves #3233

**Comment by agrrr3** [comment]
found another little thing to cleanup in order to retrigger the build.
I hope the MacOS build gets throught this time - to me it looked perfectly good last time until it was killed

**Comment by agrrr3** [comment]
next steps are w32 UI debugging and calculating the invariants after parse.

which should be rather straightforward as we established now that the parse is completely finished after the checksum calculation.

**Comment by agrrr3** [comment]
@o01eg the macOS times out for me. is it something with this branch?

**Comment by geoffthemedio** [comment]
> @o01eg the macOS times out for me. is it something with this branch?

The latest master does the same. The last successful master OSM build took 59 min. Not sure why it can't cache some of the work...

**Comment by o01eg** [comment]
Somehow it updates entire homebrew when install ccache.

**Comment by agrrr3** [comment]
rebase to o01egs build speedup

**Comment by agrrr3** [comment]
@o01eg yikes, linux travis build broken - looks like a change of base 

> o apt package "cppcheck", but there is a snap with that name.
>
> Try "snap install cppcheck"
> 
> E: Unable to locate package libboost1.62-all-dev

**Comment by o01eg** [comment]
`sudo apt update` looks missing.

**Comment by agrrr3** [comment]
Can we merge this?

**Comment by geoffthemedio** [comment]
could you rebase so the test builds rerun off the latest master?

**Pull request merged**